# 데이터 사전처리(Preprocessing)
- 머신러닝 등 데이터 분석의 정확도는 분석 데이터의 품질에 좌우된다.
- 데이터의 품질을 높이기 위해서는 누락 데이터, 중복 데이터 등 오류를 수정하고 분석목적에 맞게 변형하는 과정이 필요
- 수집한 데이터를 분석 목적에 적합하도록 하는 작업을 사전 처리(Prepreocessing)이라고 한다.

### 누락 데이터(NaN: Not a Number)
- 데이터를 파일로 입력할때 빠트리거나 파일 형식을 변환하는 과정에서 데이터가 손실되는 경우가 발생
- 머신러닝 분석 모형에 데이터를 입력하기 전에 반드시 누락 데이터를 제거하고 다른 적절한 값으로 대체 하여야 한다.
- 누락 데이터가 많아지면 데이터의 품질이 떨어지고 머신러닝 분석 알고리즘을 왜곡하는 현상이 발생한다.

In [76]:
import seaborn as sns

In [77]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


> deck에는 203개의 유효한 범주형 데이터만 있고 나머지 688개가 NaN

In [79]:
# deck열의 NaN갯수 게산하기
df['deck'].value_counts(dropna=False)

NaN    688
C       59
B       47
D       33
E       32
A       15
F       13
G        4
Name: deck, dtype: int64

# 누락 데이터 찾기

In [80]:
# NaN찾기
print(df.head().isna())

# NaN이 아닌것 찾기
print(df.head().notnull())

   survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0     False   False  False  False  False  False  False     False  False   
1     False   False  False  False  False  False  False     False  False   
2     False   False  False  False  False  False  False     False  False   
3     False   False  False  False  False  False  False     False  False   
4     False   False  False  False  False  False  False     False  False   

     who  adult_male   deck  embark_town  alive  alone  
0  False       False   True        False  False  False  
1  False       False  False        False  False  False  
2  False       False   True        False  False  False  
3  False       False  False        False  False  False  
4  False       False   True        False  False  False  
   survived  pclass   sex   age  sibsp  parch  fare  embarked  class   who  \
0      True    True  True  True   True   True  True      True   True  True   
1      True    True  True  True   True   True  True

In [81]:
# 누락 데이터 갯수 확인
df.head().isna().sum(axis=0)

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           3
embark_town    0
alive          0
alone          0
dtype: int64

In [82]:
# 누락 데이터 제거
missing_df = df.isnull()
missing_df

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0       False   False  False  False  False  False  False     False  False   
1       False   False  False  False  False  False  False     False  False   
2       False   False  False  False  False  False  False     False  False   
3       False   False  False  False  False  False  False     False  False   
4       False   False  False  False  False  False  False     False  False   
..        ...     ...    ...    ...    ...    ...    ...       ...    ...   
886     False   False  False  False  False  False  False     False  False   
887     False   False  False  False  False  False  False     False  False   
888     False   False  False   True  False  False  False     False  False   
889     False   False  False  False  False  False  False     False  False   
890     False   False  False  False  False  False  False     False  False   

       who  adult_male   deck  embark_town  alive  alone  
0    False       False   True        False  False  False  
1    False       False  False        False  False  False  
2    False       False   True        False  False  False  
3    False       False  False        False  False  False  
4    False       False   True        False  False  False  
..     ...         ...    ...          ...    ...    ...  
886  False       False   True        False  False  False  
887  False       False  False        False  False  False  
888  False       False   True        False  False  False  
889  False       False  False        False  False  False  
890  False       False   True        False  False  False  

[891 rows x 15 columns]

In [83]:
for col in missing_df.columns:
    missing_count = missing_df[col].value_counts()
    try:
        print(col,missing_count[True])
    except:
        print(col, 0)


survived 0
pclass 0
sex 0
age 177
sibsp 0
parch 0
fare 0
embarked 2
class 0
who 0
adult_male 0
deck 688
embark_town 2
alive 0
alone 0


In [84]:
# NaN값이 500개 이상인 열을 모두 삭제 
df_thresh = df.dropna(axis=1,thresh=500)
df_thresh.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')

In [85]:
# age 열에 나이 데이터가 없는 모든 행을 삭제
df_thresh_age = df_thresh.dropna(subset=['age'],how='any',axis=0)
len(df_thresh_age)

714

---
### 누락 데이터의 치환 
- 누락 데이터를 무작정 삭제해 버리면 어렵게 수집한 데이터를 활용하지 못한다.
- 데이터중 일부가 누락되어 있더라도 나머지 데이터를 최대한 살려서 데이터 분석에 활용하는 경우에 좋은 결과를 얻는 경우가 많다.

In [86]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [87]:
df['age'].head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: age, dtype: float64

In [88]:
df['age'].fillna(df['age'].mean(),inplace=True)

In [89]:
df

survived  pclass     sex        age  sibsp  parch     fare embarked  \
0           0       3    male  22.000000      1      0   7.2500        S   
1           1       1  female  38.000000      1      0  71.2833        C   
2           1       3  female  26.000000      0      0   7.9250        S   
3           1       1  female  35.000000      1      0  53.1000        S   
4           0       3    male  35.000000      0      0   8.0500        S   
..        ...     ...     ...        ...    ...    ...      ...      ...   
886         0       2    male  27.000000      0      0  13.0000        S   
887         1       1  female  19.000000      0      0  30.0000        S   
888         0       3  female  29.699118      1      2  23.4500        S   
889         1       1    male  26.000000      0      0  30.0000        C   
890         0       3    male  32.000000      0      0   7.7500        Q   

      class    who  adult_male deck  embark_town alive  alone  
0     Third    man        True  NaN  Southampton    no  False  
1     First  woman       False    C    Cherbourg   yes  False  
2     Third  woman       False  NaN  Southampton   yes   True  
3     First  woman       False    C  Southampton   yes  False  
4     Third    man        True  NaN  Southampton    no   True  
..      ...    ...         ...  ...          ...   ...    ...  
886  Second    man        True  NaN  Southampton    no   True  
887   First  woman       False    B  Southampton   yes   True  
888   Third  woman       False  NaN  Southampton    no  False  
889   First    man        True    C    Cherbourg   yes   True  
890   Third    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [90]:
# embark_town 829행의 데이터 출력
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

In [91]:
# embark_town 열의 승선도시 중에 가장 많이 출현한 값으로 치환하기
most_freq = df['embark_town'].value_counts().idxmax()
df['embark_town'].fillna(most_freq,inplace=True)
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829    Southampton
Name: embark_town, dtype: object

### 앞뒤의 유사성 데이터로 값을 치환 


In [92]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [93]:
# embark_town 열의 NaN 값을 바로 앞에 있는 828행의 값으로 변경하기
df['embark_town'].fillna(method='bfill',inplace=True)
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829      Cherbourg
Name: embark_town, dtype: object

---
# 중복 데이터 처리
- 동일한 대상이 중복으로 존재하는 것은 분석 결과의 왜곡을 초래한다.

In [94]:
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b'],
                  'c2':[1, 1, 1, 2, 2],
                  'c3':[1, 1, 2, 2, 2]})

---
# 데이터표준화
- 여러곳에서 수집한 데이터는 서로 단위가 섞여있거나 대상을 다른 형식으로 표현하는 경우가 많다.
- 대상을 표현하는 방법에 차이가 있으면 분석의 정확도는 현저히 낮아진다.
- 데이터 포맷을 일관성있게 하는 작업이 표준화이다.

In [107]:
df = pd.read_csv("../Data/auto-mpg.csv")
#df.columns = ['mpg','cylinders','displacement','horsepower','weight',
#              'acceleration','model year','origin','name']
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [100]:
df.to_csv("../Data/auto-mpg.csv",index=False)

# 단위 변환
mpg(mile per gallon) -> kpl(kilometer per liter) 로 변확    
1 mile = 1.60934 km    
l gallon = 3.78541 liter    

In [110]:
df['kpl'] = round(df['mpg']*(1.60934/3.78541),2)
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0      130.0  3504.0          12.0   
1    15.0          8         350.0      165.0  3693.0          11.5   
2    18.0          8         318.0      150.0  3436.0          11.0   
3    16.0          8         304.0      150.0  3433.0          12.0   
4    17.0          8         302.0      140.0  3449.0          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0      86.00  2790.0          15.6   
394  44.0          4          97.0      52.00  2130.0          24.6   
395  32.0          4         135.0      84.00  2295.0          11.6   
396  28.0          4         120.0      79.00  2625.0          18.6   
397  31.0          4         119.0      82.00  2720.0          19.4   

     model year  origin                       name    kpl  
0            70       1  chevrolet chevelle malibu   7.65  
1            70       1          buick skylark 320   6.38  
2            70       1         plymouth satellite   7.65  
3            70       1              amc rebel sst   6.80  
4            70       1                ford torino   7.23  
..          ...     ...                        ...    ...  
393          82       1            ford mustang gl  11.48  
394          82       2                  vw pickup  18.71  
395          82       1              dodge rampage  13.60  
396          82       1                ford ranger  11.90  
397          82       1                 chevy s-10  13.18  

[398 rows x 10 columns]

# 자료형변환


In [112]:
# 각 열의 자료형 확인
df.info()
df.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    float64
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   name          398 non-null    object 
 9   kpl           398 non-null    float64
dtypes: float64(5), int64(3), object(2)
memory usage: 31.2+ KB


mpg             float64
cylinders         int64
displacement    float64
horsepower       object
weight          float64
acceleration    float64
model year        int64
origin            int64
name             object
kpl             float64
dtype: object

> horsepower 열이 object 자료형이다. 엔진 출력을 나타내는 열이므로 숫자형으로 변환해 주는 것이 좋다.    
> 데이터를 보면 숫자처럼 보이는데 자료형이 다른것으로 나오는 것을 보면 데이터 중에 다른 자료가 있는 것이다.

In [125]:
df['horsepower'].unique()
print(len(df.horsepower))
df[(df['horsepower']=='?')]

398


mpg  cylinders  displacement horsepower  weight  acceleration  \
32   25.0          4          98.0          ?  2046.0          19.0   
126  21.0          6         200.0          ?  2875.0          17.0   
330  40.9          4          85.0          ?  1835.0          17.3   
336  23.6          4         140.0          ?  2905.0          14.3   
354  34.5          4         100.0          ?  2320.0          15.8   
374  23.0          4         151.0          ?  3035.0          20.5   

     model year  origin                  name    kpl  
32           71       1            ford pinto  10.63  
126          74       1         ford maverick   8.93  
330          80       2  renault lecar deluxe  17.39  
336          80       1    ford mustang cobra  10.03  
354          81       2           renault 18i  14.67  
374          82       1        amc concord dl   9.78

In [126]:
# 누락 데이터(?)를 삭제
# ? 를 np.nan으로 변경
import numpy as np

In [127]:
df['horsepower'].replace('?',np.nan,inplace=True)
df['horsepower'].unique()

array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', nan, '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

In [129]:
# 누락 데이터 행 삭제
df.dropna(subset=['horsepower'],axis=0, inplace=True)
len(df['horsepower'])

392

In [138]:
df['horsepower'] = df['horsepower'].astype('float')
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    float64
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   name          392 non-null    object 
 9   kpl           392 non-null    float64
dtypes: float64(6), int64(3), object(1)
memory usage: 33.7+ KB


mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0  3504.0          12.0   
1  15.0          8         350.0       165.0  3693.0          11.5   
2  18.0          8         318.0       150.0  3436.0          11.0   
3  16.0          8         304.0       150.0  3433.0          12.0   
4  17.0          8         302.0       140.0  3449.0          10.5   

   model year  origin                       name   kpl  
0          70       1  chevrolet chevelle malibu  7.65  
1          70       1          buick skylark 320  6.38  
2          70       1         plymouth satellite  7.65  
3          70       1              amc rebel sst  6.80  
4          70       1                ford torino  7.23

In [140]:
# origin의 데이터를 문자형 데이터로 변형 (1:USA, 2:EU, 3:Japan)
df['origin'].replace((1,2,3),('USA','EU','Japan'),inplace=True)
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0  3504.0          12.0   
1  15.0          8         350.0       165.0  3693.0          11.5   
2  18.0          8         318.0       150.0  3436.0          11.0   
3  16.0          8         304.0       150.0  3433.0          12.0   
4  17.0          8         302.0       140.0  3449.0          10.5   

   model year origin                       name   kpl  
0          70    USA  chevrolet chevelle malibu  7.65  
1          70    USA          buick skylark 320  6.38  
2          70    USA         plymouth satellite  7.65  
3          70    USA              amc rebel sst  6.80  
4          70    USA                ford torino  7.23

In [142]:
# origin 열의 문자형 자료형을 범주형으로 변환
df['origin'] = df['origin'].astype('category')
df['origin'].dtypes

CategoricalDtype(categories=['EU', 'Japan', 'USA'], ordered=False)

In [143]:
df['origin'] = df['origin'].astype('str')
df['origin'].dtypes

dtype('O')

In [145]:
# model year가 정수로 되어 있으나 수치계산과 관련없으므로 범주형으로 변환
df['model year'] = df['model year'].astype('category')
df['model year'].dtypes

CategoricalDtype(categories=[70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82], ordered=False)

---
# 범주형(카테고리) 데이처 처리
- 연속 데이터를 그대로 사용하기 보다는 일정한 구간(binning)으로 나눠서 분석하는것이 효율적인 경우가 많다

In [146]:
# horsepower 3개의 범위로 나누기 
count, bin_dividers = np.histogram(df['horsepower'], bins=3)
bin_dividers

array([ 46.        , 107.33333333, 168.66666667, 230.        ])

In [147]:
# 3개의 bin에 이름 지정
bin_names = ['저출력','보통출력','고출력']
df['hp_bin'] = pd.cut(
    x=df['horsepower'],
    bins=bin_dividers,
    labels = bin_names,
    include_lowest = True #첫 경계값 포함
)

In [149]:
df[['horsepower','hp_bin']].head(25)

horsepower hp_bin
0        130.0   보통출력
1        165.0   보통출력
2        150.0   보통출력
3        150.0   보통출력
4        140.0   보통출력
5        198.0    고출력
6        220.0    고출력
7        215.0    고출력
8        225.0    고출력
9        190.0    고출력
10       170.0    고출력
11       160.0   보통출력
12       150.0   보통출력
13       225.0    고출력
14        95.0    저출력
15        95.0    저출력
16        97.0    저출력
17        85.0    저출력
18        88.0    저출력
19        46.0    저출력
20        87.0    저출력
21        90.0    저출력
22        95.0    저출력
23       113.0   보통출력
24        90.0    저출력

---
### 더미(Dummy) 변수
- 범주형 데이터를 회귀분석이나 머신러닝에 바로 사용할 수 없는 경우 사용.
- 0 또는 1로 표현

In [150]:
horsepower_dummies = pd.get_dummies(df['hp_bin'])
horsepower_dummies.head(15)

저출력  보통출력  고출력
0     0     1    0
1     0     1    0
2     0     1    0
3     0     1    0
4     0     1    0
5     0     0    1
6     0     0    1
7     0     0    1
8     0     0    1
9     0     0    1
10    0     0    1
11    0     1    0
12    0     1    0
13    0     0    1
14    1     0    0

---
### One-Hot Encoding
- 범주형 데이터를 0,1을 갖는 원핫벡터로 변환
- 결과는 선형대수학에서 정의하는 희소행렬(sparse matrix)로 변환한다.
- 희소행렬은 (행,열)좌표와 값 형태로 정리

In [151]:
from sklearn import preprocessing

In [152]:
# 전처리를 위한 encoder 객체 만들기
label_encoder = preprocessing.LabelEncoder() # label encoder 생성
onehot_encoder = preprocessing.OneHotEncoder() # one hot encoder 생성  

In [153]:
# label encoder로 문자열 범주를 숫자형 범주로 변환
onehot_labeled = label_encoder.fit_transform(df['hp_bin'].head(15))
onehot_labeled

array([1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 2])

In [154]:
# 2차원 행렬로 형태를 변경
onehot_reshape = onehot_labeled.reshape(len(onehot_labeled),1)
onehot_reshape

array([[1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [2]])

In [156]:
# 희소행렬 
onehot_fitted = onehot_encoder.fit_transform(onehot_reshape)
print(onehot_fitted)

  (0, 1)	1.0
  (1, 1)	1.0
  (2, 1)	1.0
  (3, 1)	1.0
  (4, 1)	1.0
  (5, 0)	1.0
  (6, 0)	1.0
  (7, 0)	1.0
  (8, 0)	1.0
  (9, 0)	1.0
  (10, 0)	1.0
  (11, 1)	1.0
  (12, 1)	1.0
  (13, 0)	1.0
  (14, 2)	1.0


---
### 정규화(Normalization)
- 각 열에 있는 숫자 데이터의 상대적 크기 때문에 머신러닝 분석결과가 달라 질 수 있다.
- 숫자 데이터의 상대적인 크기 차이를 제거

In [158]:
# 최대값의 절대값을 이용한 정규화
df.horsepower.describe()

count    392.000000
mean     104.469388
std       38.491160
min       46.000000
25%       75.000000
50%       93.500000
75%      126.000000
max      230.000000
Name: horsepower, dtype: float64

In [160]:
# horsepower 열의 최대값의 절대값으로 모든 데이터를 나누기 하여 저장
horsepower1 = df.horsepower / abs(df.horsepower.max())
horsepower1.describe()

count    392.000000
mean       0.454215
std        0.167353
min        0.200000
25%        0.326087
50%        0.406522
75%        0.547826
max        1.000000
Name: horsepower, dtype: float64

In [161]:
# 각 열의 데이터중에 최대값과 최소값을 뺄셈한 값으로 나누기 하여 저장
df.horsepower.describe()

count    392.000000
mean     104.469388
std       38.491160
min       46.000000
25%       75.000000
50%       93.500000
75%      126.000000
max      230.000000
Name: horsepower, dtype: float64

In [162]:
# 각 열에서 해당열의 최소값을 뺀 값을 분자로 하고 
min_x = df.horsepower - df.horsepower.min()

# 해당 열의 최대값과 최소값의 차를 분모로 하는 수를 계산하면 가장 큰 값은 1이 된다.
min_max = df.horsepower.max() - df.horsepower.min()

horsepower2 = min_x/min_max
horsepower2.describe()

count    392.000000
mean       0.317768
std        0.209191
min        0.000000
25%        0.157609
50%        0.258152
75%        0.434783
max        1.000000
Name: horsepower, dtype: float64

---
# 시계열 데이터
- 특정한 시점을 기록하는 TimeStamp
- 두 지점 사이에 일정한 기간을 나타내는 Period

In [163]:
df = pd.read_csv("../Data/stock-data.csv")
df.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [164]:
df.isna().sum()

Date      0
Close     0
Start     0
High      0
Low       0
Volume    0
dtype: int64

In [165]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    20 non-null     object
 1   Close   20 non-null     int64 
 2   Start   20 non-null     int64 
 3   High    20 non-null     int64 
 4   Low     20 non-null     int64 
 5   Volume  20 non-null     int64 
dtypes: int64(5), object(1)
memory usage: 1.1+ KB


In [166]:
df['new_Date'] = pd.to_datetime(df['Date'])
df.head()

Date  Close  Start   High    Low  Volume   new_Date
0  2018-07-02  10100  10850  10900  10000  137977 2018-07-02
1  2018-06-29  10700  10550  10900   9990  170253 2018-06-29
2  2018-06-28  10400  10900  10950  10150  155769 2018-06-28
3  2018-06-27  10900  10800  11050  10500  133548 2018-06-27
4  2018-06-26  10800  10900  11000  10700   63039 2018-06-26

In [167]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Date      20 non-null     object        
 1   Close     20 non-null     int64         
 2   Start     20 non-null     int64         
 3   High      20 non-null     int64         
 4   Low       20 non-null     int64         
 5   Volume    20 non-null     int64         
 6   new_Date  20 non-null     datetime64[ns]
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 1.2+ KB


In [171]:
df.set_index('new_Date', inplace=True)
df.head()

Date  Close  Start   High    Low  Volume
new_Date                                                  
2018-07-02  2018-07-02  10100  10850  10900  10000  137977
2018-06-29  2018-06-29  10700  10550  10900   9990  170253
2018-06-28  2018-06-28  10400  10900  10950  10150  155769
2018-06-27  2018-06-27  10900  10800  11050  10500  133548
2018-06-26  2018-06-26  10800  10900  11000  10700   63039

In [175]:
df.drop('Date',axis=1, inplace=True)
df.head()

Close  Start   High    Low  Volume
new_Date                                      
2018-07-02  10100  10850  10900  10000  137977
2018-06-29  10700  10550  10900   9990  170253
2018-06-28  10400  10900  10950  10150  155769
2018-06-27  10900  10800  11050  10500  133548
2018-06-26  10800  10900  11000  10700   63039

In [176]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 20 entries, 2018-07-02 to 2018-06-01
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Close   20 non-null     int64
 1   Start   20 non-null     int64
 2   High    20 non-null     int64
 3   Low     20 non-null     int64
 4   Volume  20 non-null     int64
dtypes: int64(5)
memory usage: 1.5 KB


### Timestamp를 Period로 변환

In [177]:
# 날짜 형식의 문자열로 구성되는 리스트 정의
dates = ['2019-01-01','2020-03-01','2021-06-01']

In [178]:
ts_dates = pd.to_datetime(dates)
ts_dates

DatetimeIndex(['2019-01-01', '2020-03-01', '2021-06-01'], dtype='datetime64[ns]', freq=None)

In [180]:
pr_day = ts_dates.to_period(freq = 'D')
print(pr_day)
pr_month = ts_dates.to_period(freq = 'M')
print(pr_month)
pr_year = ts_dates.to_period(freq = 'Y')
print(pr_year)

PeriodIndex(['2019-01-01', '2020-03-01', '2021-06-01'], dtype='period[D]')
PeriodIndex(['2019-01', '2020-03', '2021-06'], dtype='period[M]')
PeriodIndex(['2019', '2020', '2021'], dtype='period[A-DEC]')


> freq Option   
> D (Day: 1일)   
> W (Week:1주)   
> M (Month end: 월말)   
> MS (Month begin: 월초)   
> Q (Quarter end : 분기말)   
> QS (Quarter begin: 분기초)   
> A (year end : 연말)   
> AS (year begin: 연초)   
> B (Business day : 휴일 제외)   
> H (hour : 1시간)   
> T (minute :1분)   
> S (second: 1초)

# 시계열 데이터 만들기

In [183]:
# Timestamp의 배열 만들기 : 월 간격, 월의 시작일 기준
ts_ms = pd.date_range(
    start='2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods = 6, # 생성할 Timestampe의 갯수
    freq='MS', # 시간 간격
    tz = 'Asia/Seoul' # 시간대
)
ts_ms

DatetimeIndex(['2022-01-01 00:00:00+09:00', '2022-02-01 00:00:00+09:00',
               '2022-03-01 00:00:00+09:00', '2022-04-01 00:00:00+09:00',
               '2022-05-01 00:00:00+09:00', '2022-06-01 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='MS')

In [185]:
# Timestamp의 배열 만들기 : 월 간격, 월의 마지막날 기준
ts_me = pd.date_range(
    start='2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods = 6, # 생성할 Timestampe의 갯수
    freq='M', # 시간 간격
    tz = 'Asia/Seoul' # 시간대
)
ts_me

DatetimeIndex(['2022-01-31 00:00:00+09:00', '2022-02-28 00:00:00+09:00',
               '2022-03-31 00:00:00+09:00', '2022-04-30 00:00:00+09:00',
               '2022-05-31 00:00:00+09:00', '2022-06-30 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='M')

In [186]:
# Timestamp의 배열 만들기 : 분기 간격, 월의 마지막날 기준
ts_q = pd.date_range(
    start='2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods = 6, # 생성할 Timestampe의 갯수
    freq='Q', # 시간 간격
    tz = 'Asia/Seoul' # 시간대
)
ts_q

DatetimeIndex(['2022-03-31 00:00:00+09:00', '2022-06-30 00:00:00+09:00',
               '2022-09-30 00:00:00+09:00', '2022-12-31 00:00:00+09:00',
               '2023-03-31 00:00:00+09:00', '2023-06-30 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='Q-DEC')

---
# Period 배열


In [187]:
# 1개월 길이
pr_m = pd.period_range(
    start='2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods = 3, # 생성할 Period의 갯수
    freq='M', # 시간 간격
)
pr_m

PeriodIndex(['2022-01', '2022-02', '2022-03'], dtype='period[M]')

In [188]:
# 2시간 길이
pr_2h = pd.period_range(
    start='2022-01-01', # 날짜 범위의 시작
    end = None, # 날짜 범위의 끝
    periods = 3, # 생성할 Period의 갯수
    freq='2H', # 시간 간격
)
pr_2h

PeriodIndex(['2022-01-01 00:00', '2022-01-01 02:00', '2022-01-01 04:00'], dtype='period[2H]')

In [189]:
df.reset_index(inplace=True)
df.head()

new_Date  Close  Start   High    Low  Volume
0 2018-07-02  10100  10850  10900  10000  137977
1 2018-06-29  10700  10550  10900   9990  170253
2 2018-06-28  10400  10900  10950  10150  155769
3 2018-06-27  10900  10800  11050  10500  133548
4 2018-06-26  10800  10900  11000  10700   63039

In [191]:
df['Year'] = df['new_Date'].dt.to_period(freq = 'Y')
df['Month'] = df['new_Date'].dt.to_period(freq = 'M')
df['Date'] = df['new_Date'].dt.to_period(freq = 'D')

In [193]:
df['Date_yr'] = df['new_Date'].dt.to_period(freq='A')
df['Date_m'] = df['new_Date'].dt.to_period(freq='M')
df.head()

new_Date  Close  Start   High    Low  Volume  Year    Month        Date  \
0 2018-07-02  10100  10850  10900  10000  137977  2018  2018-07  2018-07-02   
1 2018-06-29  10700  10550  10900   9990  170253  2018  2018-06  2018-06-29   
2 2018-06-28  10400  10900  10950  10150  155769  2018  2018-06  2018-06-28   
3 2018-06-27  10900  10800  11050  10500  133548  2018  2018-06  2018-06-27   
4 2018-06-26  10800  10900  11000  10700   63039  2018  2018-06  2018-06-26   

  Date_yr   Date_m  
0    2018  2018-07  
1    2018  2018-06  
2    2018  2018-06  
3    2018  2018-06  
4    2018  2018-06

In [194]:
df.set_index('new_Date',inplace=True)
df.head()

Close  Start   High    Low  Volume  Year    Month        Date  \
new_Date                                                                    
2018-07-02  10100  10850  10900  10000  137977  2018  2018-07  2018-07-02   
2018-06-29  10700  10550  10900   9990  170253  2018  2018-06  2018-06-29   
2018-06-28  10400  10900  10950  10150  155769  2018  2018-06  2018-06-28   
2018-06-27  10900  10800  11050  10500  133548  2018  2018-06  2018-06-27   
2018-06-26  10800  10900  11000  10700   63039  2018  2018-06  2018-06-26   

           Date_yr   Date_m  
new_Date                     
2018-07-02    2018  2018-07  
2018-06-29    2018  2018-06  
2018-06-28    2018  2018-06  
2018-06-27    2018  2018-06  
2018-06-26    2018  2018-06

In [197]:
# 날짜 인덱스만 이용하여 데이터 선택하기
df_y = df.loc['2018-07','Start':'High']
df_y

Start   High
new_Date                
2018-07-02  10850  10900

In [198]:
# 시간 간격 계산, 최근 180일 ~ 189일 사이의 값들만 선택하기
today = pd.to_datetime('2018-12-25')
df['time_delta'] = today - df.index
df.set_index('time_delta',inplace=True)
df_180 = df['180 days':'189 days']
df_180

Close  Start   High    Low  Volume  Year    Month        Date  \
time_delta                                                                  
180 days    10400  10900  10950  10150  155769  2018  2018-06  2018-06-28   
181 days    10900  10800  11050  10500  133548  2018  2018-06  2018-06-27   
182 days    10800  10900  11000  10700   63039  2018  2018-06  2018-06-26   
183 days    11150  11400  11450  11000   55519  2018  2018-06  2018-06-25   
186 days    11300  11250  11450  10750  134805  2018  2018-06  2018-06-22   
187 days    11200  11350  11750  11200  133002  2018  2018-06  2018-06-21   
188 days    11550  11200  11600  10900  308596  2018  2018-06  2018-06-20   
189 days    11300  11850  11950  11300  180656  2018  2018-06  2018-06-19   

           Date_yr   Date_m  
time_delta                   
180 days      2018  2018-06  
181 days      2018  2018-06  
182 days      2018  2018-06  
183 days      2018  2018-06  
186 days      2018  2018-06  
187 days      2018  2018-06  
188 days      2018  2018-06  
189 days      2018  2018-06